### PR TITLE
Initial suite of tests

### PR DESCRIPTION
**Pull request by eschrom**
Many tests are included here for data cleaning, uptake data manipulating, and model building. Writing tests inspired some moderate reorganizations/fixes in __init__.py. The tests for model building are a bit sparse, because this portion of the code will be significantly reworked (requiring new tests) when numpyro is introduced. Still, it would be good to get some legitimate testing infrastructure incorporated.

### CODE REVIEW

File: tests/test_model_building.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+def test_extract_standards(frame):
+    """
+    Make a dictionary with as many keys as provided variables,
+    and two values per sub-dictionary
+    """
+    output = iup.LinearIncidentUptakeModel.extract_standards(
+        frame, ("estimate", "elapsed")
+    )
+
+    assert len(output) == 2
```
Comment (swo) [inline tests/test_model_building.py:74]: it's a minor thing, but I might be more specific here. Can you you write out what you expect `output` to be in this situation, like the exact data frame? That makes it easier for me as the code reviewer to be 100% sure about what you want this function to be doing

File: tests/test_model_building.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+def test_fit(frame):
+    """
+    Model should fit a line to two points, giving a 'perfect' fit
+    """
+
+    output = iup.LinearIncidentUptakeModel()
+    output = output.fit(frame, ("geography",))
+
+    assert output.model.score(output.x, output.y) == 1.0
```
Comment (swo) [inline tests/test_model_building.py:86]: would be nice to also check that the coefficient(s) in the model is precisely what you think it was

File: tests/test_model_building.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    start_date = dt.date(2020, 2, 1)
+    end_date = dt.date(2020, 2, 29)
+    interval = "7d"
+    group_cols = None
+
+    output = iup.LinearIncidentUptakeModel.build_scaffold(
+        start, start_date, end_date, interval, group_cols
+    )
+
+    assert output.shape[0] == 5
```
Comment (swo) [inline tests/test_model_building.py:111]: you could write this as `output.shape == (5, 7)`

and this is another place that saying the exact data frame you expect would be good

File: tests/test_model_building.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    output = iup.LinearIncidentUptakeModel.build_scaffold(
+        start, start_date, end_date, interval, group_cols
+    )
+
+    assert output.shape[0] == 10
+    assert output.shape[1] == 8
+
+
+def test_project_sequentially():
```
Comment (swo) [inline tests/test_model_building.py:141]: love this test. are there any other trivial or edge cases that are easy to test?

File: iup/__init__.py [hunk 5]
```diff
@@ -735,81 +793,163 @@ def predict(
...
+        Projections are made separately by group, if grouping factors exist.
+        This function handles one group at a time.
+
+        This does not yet incorporate uncertainty, but it should in the future.
+        """
+        proj = np.zeros(len(elapsed) + 1)
+
+        proj[0] = start
+
+        for i in range(proj.shape[0] - 1):
```
[the remark this answers is not in the document]
Comment (swo) [reply iup/__init__.py:864]: this is also something that I think you could reframe as an accumulation; but that's mostly for kicks I'd say

**Review by swo — APPROVED** [review]
A number of suggestions and some requests for documentation, but nothing critical or blocking. I think a lot of the stuff that I'm thinking should be tweaked are things that we'll develop joint understanding of, when we implement a second model.

File: iup/__init__.py [hunk 6]
```diff
@@ -225,6 +225,44 @@ def trim_outlier_intervals(
 
         return IncidentUptakeData(df)
 
+    def expand_implicit_columns(self, group_cols: tuple[str,] | None) -> Self:
```
[the remark this answers is not in the document]
Comment (eschrom) [reply iup/__init__.py:228]: "Augment" is actually the word I found myself tending toward when writing documentation, so I will happily change it.

File: iup/__init__.py [hunk 7]
```diff
@@ -225,6 +225,44 @@ def trim_outlier_intervals(
...
+        Details
+        Extra columns are added to the incident uptake data:
+        - disease season that each date belongs to
+        - interval of time in days between each successive date
+        - number of days elapsed between rollout and each date
+        - daily-average uptake in the interval preceding each date
+        - daily-average uptake in the interval preceding the previous date
+        """
+        self = (
+            self.with_columns(
```
[the remark this answers is not in the document]
Comment (eschrom) [reply iup/__init__.py:251]: I will add this to issue #42.

File: iup/__init__.py [hunk 8]
```diff
@@ -318,7 +356,7 @@ def parse_nis(
         format of the dates in the NIS date column
     rollout: dt.date
         date of rollout
-    filters: dict
+    filters: dict | None
```
[the remark this answers is not in the document]
Comment (eschrom) [reply iup/__init__.py:359]: Good to know! I will open a new issue #55 to remind me to make type hinting as precise as possible throughout.

File: iup/__init__.py [hunk 9]
```diff
@@ -406,7 +445,7 @@ def select_columns(
         name of the NIS column for the uptake estimate (population %)
     date_col: str
         name of the NIS column for the date
-    group_cols: dict
+    group_cols: dict | None
```
[the remark this answers is not in the document]
Comment (eschrom) [reply iup/__init__.py:448]: While it might be a touch confusing when inspecting `select_columns` for the first time, I think it makes a really intuitive interface for the user in the config.yaml. You just specify a nested dictionary of: `{"what_the_grouping_column_is_called" : "what_I_want_the_grouping_column_to_be_called"}`. This makes sure the user enters the same number of actual vs. desired grouping column names, whereas a list for each could be different lengths (raising errors downstream). 

The 3rd example you raise shouldn't be an issue, because this is only for grouping columns. While `{"raw_point_estimate": "estimate"}` would indeed rename a column, it would also cause errors downstream because the point estimate would be erroneously used as a grouping factor.

File: iup/__init__.py [hunk 10]
```diff
@@ -516,37 +550,47 @@ def standardize(x, mn=None, sd=None):
...
     """
-    if mn is not None:
-        return (x - mn) / sd
+    if type(x) is pl.Expr:
+        if mn is not None:
+            return (x - mn) / sd
+        else:
+            return (
+                pl.when(x.drop_nulls().n_unique() == 1)
+                .then(x * 0.0)
```
[the remark this answers is not in the document]
Comment (eschrom) [reply iup/__init__.py:570]: Change made and documentation added!

File: iup/__init__.py [hunk 11]
```diff
@@ -735,81 +793,163 @@ def predict(
             )
             .alias("date")
             .to_frame()
+            .with_columns(estimate=pl.lit(0.0))
```
[the remark this answers is not in the document]
Comment (eschrom) [reply iup/__init__.py:796]: The next step is to `augment_implicit_columns` which makes the "daily" column by "estimate" / "interval" and then shifts this to make the "previous" column. Null would then populate these two columns, as well. In the "previous" column, null should be reserved for the first date, which doesn't have a previous value, not the entire column.

File: iup/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eschrom) [reply iup/__init__.py:864]: Detail added!

File: tests/test_data_cleaning.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+def test_apply_filters_handles_no_filters(frame):
+    """
+    If no filters are given to apply_filters, the whole frame is returned.
+    """
+    filters = None
+
+    output = iup.apply_filters(frame, filters)
+
+    assert output.shape[0] == 3
```
[the remark this answers is not in the document]
Comment (eschrom) [reply tests/test_data_cleaning.py:44]: Or how about `assert output.equals(frame)`?

File: tests/test_data_cleaning.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+def test_extract_group_names_handles_unmatched_values():
+    """
+    If a list of group name dictionaries with mismatched values is given, an error is raised.
+    """
+    group_cols = [
+        {"geography": "region", "indicator": "outcome"},
+        {"geography": "area", "indicator": "outcome"},
+    ]
+
+    with pytest.raises(AssertionError):
```
[the remark this answers is not in the document]
Comment (eschrom) [reply tests/test_data_cleaning.py:152]: Good point! I've moved this to issue #56, because I suspect we need more tests confirming that functions fail when expected with an appropriate error.

File: tests/test_model_building.py (hunk 1 above)
Comment (eschrom) [reply tests/test_model_building.py:74]: Sure, done.

File: tests/test_model_building.py (hunk 2 above)
Comment (eschrom) [reply tests/test_model_building.py:86]: Good idea. I as we add more types of models, I think it would be good to choose a single fixture data frame for testing that has the right edge cases and known outputs to thoroughly test all models.

File: tests/test_model_building.py (hunk 4 above)
Comment (eschrom) [reply tests/test_model_building.py:141]: Likely so - I'll give it some thought but merge what I have so far.
